python/ray/dag/tests/experimental/test_execution_schedule.py

```diff
@@ -0,0 +1,1017 @@
+# coding: utf-8
+import os
+import sys
+
+import pytest
+
+import ray
+import ray.cluster_utils
+from ray.experimental.channel.torch_tensor_type import TorchTensorType
+from ray.tests.conftest import *  # noqa
+from ray.dag import InputNode, MultiOutputNode, ClassMethodNode
+from ray.dag.dag_node_operation import (
+    DAGNodeOperationType,
+    DAGOperationGraphNode,
+    DAGNodeOperation,
+)
+from ray.dag.compiled_dag_node import _select_next_nodes, CompiledDAG, CompiledTask
+import torch
+from typing import List, Dict, Tuple
+from dataclasses import dataclass, field
+from collections import deque, defaultdict
+from ray.actor import ActorHandle
+
+if sys.platform != "linux" and sys.platform != "darwin":
+    pytest.skip("Skipping, requires Linux or Mac.", allow_module_level=True)
+
+USE_GPU = bool(os.environ.get("RAY_PYTEST_USE_GPU", 0))
+
+
+@dataclass
+class PipelineConfig:
+    """
+    pp_size: Number of pipeline parallel workers.
+    num_micro_batches: Number of micro-batches.
+    """
+
+    pp_size: int
+    num_micro_batches: int
+
+
+@dataclass
+class PipelineUnit:
+    """
+    op: Operation type (FWD or BWD).
+    pp_rank: Pipeline parallel rank.
+    batch_id: Batch ID.
+    uid: Unique ID for the pipeline unit.
+    """
+
+    op: str
+    pp_rank: int
+    batch_id: int
+    uid: str = field(init=False, repr=False)
+
+    def __post_init__(self):
+        self.uid = f"{self.op}_rank-{self.pp_rank}_batch-{self.batch_id}"
+
+    def __repr__(self) -> str:
+        return self.uid
+
+
+def generate_1f1b_schedule(config: PipelineConfig) -> List[List[PipelineUnit]]:
+    """
+    Args:
+        config: Pipeline configuration.
+    Returns:
+        schedule: List of pipeline units for 1F1B pipeline parallelism. Each
+            inner list represents the schedule for a pipeline parallel worker.
+    """
+    pp_size = config.pp_size
+    num_micro_batches = config.num_micro_batches
+
+    schedule = []
+    for pp_rank in range(config.pp_size):
+        warm_up_batches = pp_size - pp_rank
+        main_1f1b_batches = num_micro_batches - warm_up_batches
+        cool_down_batches = num_micro_batches - main_1f1b_batches
+
+        rank_schedule = []
+        bwd_batch_id = fwd_batch_id = 0
+
+        for _ in range(warm_up_batches):
+            rank_schedule.append(PipelineUnit("FWD", pp_rank, fwd_batch_id))
+            fwd_batch_id += 1
+
+        for _ in range(main_1f1b_batches):
+            rank_schedule.append(PipelineUnit("BWD", pp_rank, bwd_batch_id))
+            bwd_batch_id += 1
+            rank_schedule.append(PipelineUnit("FWD", pp_rank, fwd_batch_id))
+            fwd_batch_id += 1
+
+        for _ in range(cool_down_batches):
+            rank_schedule.append(PipelineUnit("BWD", pp_rank, bwd_batch_id))
+            bwd_batch_id += 1
+        schedule.append(rank_schedule)
+    return schedule
+
+
+class PipelineModel:
```

Comment:
Hmm this code is a bit more complicated than I expected it to be. Why not just build the pipeline schedule directly into the DAG and skip the intermediate `schedule: List[List[PipelineUnit]]` interface? I.e. everywhere we are running `rank_schedule.append(PipelineUnit))`, just replace it with `self.actors[rank].fwd.bind(...)`?

Reply:
cc @woshiyyya

Reply:
@stephanie-wang 

`self.actors[rank].fwd.bind(parent_node)`

We cannot bind `parent_node` on the fly while we are defining the rank schedule, because the `parent_node` has not been initialized yet.

![PP@2x (2)](https://github.com/user-attachments/assets/58877f87-2a56-4fa4-b16a-a508098683d1)

For example, to define node A, we have to bind node B (data dependency). However node B won't be initialized before we define node C,D,E,F,...(control dependency).

Therefore, the only choice for me was to define the dag node according to topological order.

Reply:
I think the core issue here is there's no way to explicitly define control dependency. Users have to define the node in the correct order to implicitly achieve the desired per rank schedule.

Reply:
I'm not sure I completely understand the problem. Here is a [script](https://gist.github.com/stephanie-wang/7a80d96e55b1245349588de165190f7d) for 1f1b, can you check if this is different from the test here?

Although the script I wrote hangs with aDAGs even without NCCL, which seems to be a separate bug... will file an issue for this.

Reply:
(filed an issue here: https://github.com/ray-project/ray/issues/47039)

Reply:
<img width="1119" alt="Screenshot 2024-08-08 at 11 40 02 PM" src="https://github.com/user-attachments/assets/2e7e1280-bec3-4df8-b5ab-107427c5b612">

I printed the generated schedule and it seems not equivalent to 1F1B.

```
F0 F1 F2 F3 B0 F4 B1 F5 B2 F6 B3 F7 B4 B5 B6 B7
F0 F1 F2 F3 B0 B1 F4 B2 F5 B3 F6 B4 F7 B5 B6 B7
F0 F1 F2 F3 B0 B1 B2 F4 B3 F5 B4 F6 B5 F7 B6 B7
F0 F1 F2 F3 B0 B1 B2 B3 F4 B4 F5 B5 F6 B6 F7 B7
```

Reply:
I think this can be fixed by changing [line 37](https://gist.github.com/stephanie-wang/7a80d96e55b1245349588de165190f7d#file-1f1b-py-L37) to 
```
fwd_counter = [NUM_LEAD_MICROBATCHES - i for i in range(NUM_WORKERS)]
```

The schedule after the fix:
```
F0 F1 F2 F3 B0 F4 B1 F5 B2 F6 B3 F7 B4 B5 B6 B7
F0 F1 F2 B0 F3 B1 F4 B2 F5 B3 F6 B4 F7 B5 B6 B7
F0 F1 B0 F2 B1 F3 B2 F4 B3 F5 B4 F6 B5 F7 B6 B7
F0 B0 F1 B1 F2 B2 F3 B3 F4 B4 F5 B5 F6 B6 F7 B7
```

Reply:
Ah nice catch! I'll update the script.

Reply:
Updated https://github.com/ray-project/ray/pull/46911/commits/e006cd8c8fed701a4da6ed38c353393147998cb6. Thanks @stephanie-wang and @woshiyyya!